[dynamo] fix bug where non-recursive disable modifies the original function

williamwen42 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #148896

Fixes https://github.com/pytorch/pytorch/issues/148787.

We fix this by:
- Wrapping the original function instead of directly modifying it
- When we detect that the previous frame is the non-recursive disable wrapper, then skip tracing this frame (non-recursive disable wrapper will always be skipped, so that frame will be present in the traceback)l

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

williamwen42 committed:
[dynamo] fix bug where non-recursive disable modifies the original function

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] fix bug where non-recursive disable modifies the original function"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] fix bug where non-recursive disable modifies the original function"


Fixes https://github.com/pytorch/pytorch/issues/148787.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] fix bug where non-recursive disable modifies the original function"


Fixes https://github.com/pytorch/pytorch/issues/148787.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] fix bug where non-recursive disable modifies the original function"


Fixes https://github.com/pytorch/pytorch/issues/148787.

We fix this by:
- Wrapping the original function instead of directly modifying it
- We enable convert_frame to signal to eval_frame to skip tracing the next frame evaluated (NOT the next frame traced)
- When we detect that the disable non-recursive wrapper is traced, convert_frame signals to eval_frame to skip the next frame.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jansel reviewed (changes requested):
Test failure/lint?

williamwen42 committed:
Update on "[dynamo] fix bug where non-recursive disable modifies the original function"


Fixes https://github.com/pytorch/pytorch/issues/148787.

We fix this by:
- Wrapping the original function instead of directly modifying it
- We enable convert_frame to signal to eval_frame to skip tracing the next frame evaluated (NOT the next frame traced)
- When we detect that the disable non-recursive wrapper is traced, convert_frame signals to eval_frame to skip the next frame.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] fix bug where non-recursive disable modifies the original function"


Fixes https://github.com/pytorch/pytorch/issues/148787.

We fix this by:
- Wrapping the original function instead of directly modifying it
- We enable convert_frame to signal to eval_frame to skip tracing the next frame evaluated (NOT the next frame traced)
- When we detect that the disable non-recursive wrapper is traced, convert_frame signals to eval_frame to skip the next frame.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] fix bug where non-recursive disable modifies the original function"


Fixes https://github.com/pytorch/pytorch/issues/148787.

We fix this by:
- Wrapping the original function instead of directly modifying it
- We enable convert_frame to signal to eval_frame to skip tracing the next frame evaluated (NOT the next frame traced)
- When we detect that the disable non-recursive wrapper is traced, convert_frame signals to eval_frame to skip the next frame.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] fix bug where non-recursive disable modifies the original function"


Fixes https://github.com/pytorch/pytorch/issues/148787.

We fix this by:
- Wrapping the original function instead of directly modifying it
- We enable convert_frame to signal to eval_frame to skip tracing the next frame evaluated (NOT the next frame traced)
- When we detect that the disable non-recursive wrapper is traced, convert_frame signals to eval_frame to skip the next frame.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] fix bug where non-recursive disable modifies the original function"


Fixes https://github.com/pytorch/pytorch/issues/148787.

We fix this by:
- Wrapping the original function instead of directly modifying it
- We enable convert_frame to signal to eval_frame to skip tracing the next frame evaluated (NOT the next frame traced)
- When we detect that the disable non-recursive wrapper is traced, convert_frame signals to eval_frame to skip the next frame.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] fix bug where non-recursive disable modifies the original function"


Fixes https://github.com/pytorch/pytorch/issues/148787.

We fix this by:
- Wrapping the original function instead of directly modifying it
- We enable convert_frame to signal to eval_frame to skip tracing the next frame evaluated (NOT the next frame traced)
- When we detect that the disable non-recursive wrapper is traced, convert_frame signals to eval_frame to skip the next frame.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] fix bug where non-recursive disable modifies the original function"


Fixes https://github.com/pytorch/pytorch/issues/148787.

We fix this by:
- Wrapping the original function instead of directly modifying it
- We enable convert_frame to signal to eval_frame to skip tracing the next frame evaluated (NOT the next frame traced)
- When we detect that the disable non-recursive wrapper is traced, convert_frame signals to eval_frame to skip the next frame.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jansel commented on torch/csrc/dynamo/eval_frame.c:
```diff
@@ -460,6 +460,8 @@ PyObject* dynamo_eval_custom_code(
   return result;
 }
 
+bool skip_next_frame = false;
```
Is this thread safe?  Does it need to be thread local?

jansel reviewed (changes requested):
I am worried that the `skip_next_frame` approach is:
1) Not thread safe (the next frame might be a different thread if you are unlucky).
2) Clunky and hard to reason about
3) Could result in undetected bugs if someone adds a frame call between the two frames.

I'd suggest an alternate approach:
When we call into dynamo (only at compile time, zero runtime cost) look at the frame above us on the stack (using traceback/sys) and check if the frame above is the non-recursive skip frame.  If so, return SkipFrame.

williamwen42 committed:
Update on "[dynamo] fix bug where non-recursive disable modifies the original function"


Fixes https://github.com/pytorch/pytorch/issues/148787.

We fix this by:
- Wrapping the original function instead of directly modifying it
- We enable convert_frame to signal to eval_frame to skip tracing the next frame evaluated (NOT the next frame traced)
- When we detect that the disable non-recursive wrapper is traced, convert_frame signals to eval_frame to skip the next frame.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] fix bug where non-recursive disable modifies the original function"


Fixes https://github.com/pytorch/pytorch/issues/148787.

We fix this by:
- Wrapping the original function instead of directly modifying it
- When we detect that the previous frame is the non-recursive disable wrapper, then skip tracing this frame (non-recursive disable wrapper will always be skipped, so that frame will be present in the traceback)l

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 committed:
Update on "[dynamo] fix bug where non-recursive disable modifies the original function"


Fixes https://github.com/pytorch/pytorch/issues/148787.

We fix this by:
- Wrapping the original function instead of directly modifying it
- When we detect that the previous frame is the non-recursive disable wrapper, then skip tracing this frame (non-recursive disable wrapper will always be skipped, so that frame will be present in the traceback)l

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

williamwen42 commented:
@jansel changed approach to look at traceback for the nonrecursive disable wrapper, instead of signaling eval_frame to skip the next frame

williamwen42 commented:
@pytorchbot merge
